src/main/java/org/commonjava/util/sidecar/util/BufferStreamingOutput.java

```diff
@@ -0,0 +1,86 @@
+package org.commonjava.util.sidecar.util;
+
+import io.vertx.mutiny.core.buffer.Buffer;
+import io.vertx.mutiny.ext.web.client.HttpResponse;
+import org.apache.commons.io.output.CountingOutputStream;
+import org.apache.commons.io.output.TeeOutputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.StreamingOutput;
+import javax.xml.bind.DatatypeConverter;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+
+public class BufferStreamingOutput
+                implements StreamingOutput
+{
+    private static final int bufSize = 10 * 1024 * 1024;
+
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    private HttpResponse<Buffer> response;
+
+    private Supplier<OutputStream> cacheStreamSupplier;
+
+    public BufferStreamingOutput( HttpResponse<Buffer> response )
+    {
+        this.response = response;
+    }
+
+    @Override
+    public void write( OutputStream output ) throws IOException, WebApplicationException
+    {
+        OutputStream cacheStream = null;
+        try(CountingOutputStream cout = new CountingOutputStream( output ))
+        {
+            OutputStream out = cout;
+            if ( cacheStreamSupplier != null )
+            {
+                cacheStream = cacheStreamSupplier.get();
+                if ( cacheStream != null )
+                {
+                    out = new TeeOutputStream( cacheStream, output );
+                }
+            }
+
+            Buffer buffer = response.bodyAsBuffer();
+            int total = buffer.length();
+            int transferred = 0;
+            while ( transferred < total )
+            {
+                int next = bufSize < total ? bufSize : total;
+                byte[] bytes = buffer.getBytes( transferred, next );
+                out.write( bytes );
+
+                transferred = next;
```

Comment:
I'm thinking the quit condition here for the while loop. There are two cases:
1. For small files whose size < bufSize, so there will be only one loop and can quit successfully.
2. For large files whose size > bufSize. In this case, the "total" will be the real file size(from the buffer.length), right? So that means the transferred will always be set to bufSize because of the "bufSize<total". Will this cause a dead loop as transferred will be less than total forever?